test/integration-test/src/tests/relocations.rs

```diff
@@ -0,0 +1,247 @@
+            // NOTE: these tests depend on libbpf, LIBBPF_INCLUDE must point its headers.
+            // This is set automatically by the integration-test xtask.
+            .args([
+                "-I",
+                &std::env::var("LIBBPF_INCLUDE").context("LIBBPF_INCLUDE not set")?,
+            ])
+            .arg(&ebpf_file)
+            .status()
+            .context("Failed to run clang")?
+            .success()
+            .then_some(())
+            .context("Failed to compile eBPF program")?;
+        let ebpf = std::fs::read(ebpf_file.with_extension("o"))
+            .context("Error reading compiled eBPF program")?;
+        Ok(ebpf)
+    }
+
+    /// - Generate the target BTF source with a mock main()
+    /// - Compile it with clang
+    /// - Extract the BTF with pahole
+    fn build_btf(&self) -> Result<Btf> {
+        let target_btf = self.target_btf;
+        let tmp_dir = tempfile::tempdir().context("Error making temp dir")?;
+        // BTF files can be generated and inspected with these commands:
+        // $ clang -c -g -O2 -target bpf target.c
+        // $ pahole --btf_encode_detached=target.btf -V target.o
+        // $ bpftool btf dump file ./target.btf  format c
+        let btf_file = tmp_dir.path().join("target_btf.c");
+        std::fs::write(
+            &btf_file,
+            format!(
+                r#"
+                    #include <linux/types.h>
+                    {target_btf}
+                    int main() {{ return 0; }}
+                "#
+            ),
+        )
+        .context("Writing bpf program failed")?;
+        Command::new("clang")
+            .current_dir(tmp_dir.path())
+            .args(["-c", "-g", "-O2", "-target", "bpf"])
+            .arg(&btf_file)
+            .status()
+            .context("Failed to run clang")?
+            .success()
+            .then_some(())
+            .context("Failed to compile BTF")?;
+        Command::new("pahole")
+            .current_dir(tmp_dir.path())
+            .arg("--btf_encode_detached=target.btf")
+            .arg(btf_file.with_extension("o"))
+            .status()
+            .context("Failed to run pahole")?
+            .success()
+            .then_some(())
+            .context("Failed to extract BTF")?;
+        let btf = Btf::parse_file(tmp_dir.path().join("target.btf"), Endianness::default())
+            .context("Error parsing generated BTF")?;
+        Ok(btf)
+    }
+}
+
+struct RelocationTestRunner {
+    ebpf: Vec<u8>,
+    btf: Btf,
+}
+
+impl RelocationTestRunner {
+    /// Run test and return the output value
+    fn run(&self) -> Result<u32> {
+        self.run_internal(true)
+    }
+
+    /// Run without loading btf
+    fn run_no_btf(&self) -> Result<u32> {
+        self.run_internal(false)
+    }
+
+    fn run_internal(&self, with_relocations: bool) -> Result<u32> {
+        let mut loader = BpfLoader::new();
+        if with_relocations {
+            loader.btf(Some(&self.btf));
+        }
```

Comment:
by default the loader will load BTF info from sysfs if `btf()` is invoked, so here we need an else branch that does `loader.btf(None)`